File: Labs/Algorithms4.2/mincost/mincost.cpp
#include <iostream>
#include <vector>
#include <deque>

using namespace std;

const long long INF = 1e12;

struct Edge {
    Edge(unsigned long long from, unsigned long long to, long long cap, 
    long long flow, long long price, unsigned long long rev) : 
    from(from), to(to), cap(cap), flow(flow), price(price), rev(rev) {}
    
    unsigned long long from, to;
    long long cap, flow, price;
    unsigned long long rev;
};

int main()
{
    freopen("mincost.in", "r", stdin);
    freopen("mincost.out", "w", stdout);
    
    unsigned int n, m;
    scanf("%u %u", &n, &m);
    
    vector<Edge> edges;
    vector< vector<unsigned long long> > graph(n, vector<unsigned long long>());
    
    for (size_t i = 0; i < m; ++i)
    {
        unsigned long long from, to;
        long long cap, price;
        scanf("%llu %llu %lld %lld", &from, &to, &cap, &price);
        edges.emplace_back(from - 1, to - 1, cap, 0, price, 0);
        edges.emplace_back(to - 1, from - 1, 0, 0, -price, 0);
        unsigned long long len = edges.size();
        edges[len - 1].rev = len - 2;
        edges[len - 2].rev = len - 1;
        graph[edges[len - 1].from].push_back(len - 1);
        graph[edges[len - 2].from].push_back(len - 2);
    }
    
    long long cost(0);
    
    vector<long long> d;
    vector<unsigned long long> p;
    deque<unsigned long long> q;
    vector<short> i;
    
    while (true)
    {
        d.assign(n, INF);
        p.assign(n, INF);
        i.assign(n, 0);
        d[0] = 0;
        q.push_back(0);
        
        while (!q.empty())
        {
            unsigned long long v = q.front();
            q.pop_front();
            i[v] = 1;
            for (size_t u = 0; u < graph[v].size(); ++u)
            {
                Edge & edge = edges[graph[v][u]];
                if (d[v] < INF
                && d[edge.to] > d[v] + edge.price
                && edge.flow < edge.cap)
                {
                    d[edge.to] = d[v] + edge.price;
                    p[edge.to] = graph[v][u];
                    if (i[edge.to] == 0)
                    {
                        q.push_back(edge.to);
                    }
                    else
                    {
                        if (i[edge.to] == 1)
                        {
                            q.push_front(edge.to);
                        }
                    }
                    i[edge.to] = 1;
                }
            }
        }
        
        if (d[n - 1] == INF)
        {
            break;
        }
        else
        {
            unsigned long long cur = n - 1;
            long long minF = INF;
            while (p[cur] != INF)
            {
                minF = min(minF, edges[p[cur]].cap - edges[p[cur]].flow);
                cur = edges[p[cur]].from;
            }
            cur = n - 1;
            while (p[cur] != INF)
            {
                edges[p[cur]].flow += minF;
                unsigned long long rev = edges[p[cur]].rev;
                edges[rev].flow -= minF;
                cost += minF * edges[p[cur]].price;
                cur = edges[p[cur]].from;
            }
        }
    }
    
    printf("%lld\n", cost);
    
    return 0;
}

File: Labs/Algorithms4.2/mincost/mincost.rs
use std::io::Write;
use std::io::Read;
use std::fs::File;
use std::collections::VecDeque;

struct Edge {
    from  : usize,
    to    : usize,
    cap   : i64,
    flow  : i64,
    price : i64,
    rev   : usize,
}

fn main() {
    let mut input = File::open("mincost.in").unwrap();
    let mut output = File::create("mincost.out").unwrap();
    
    let mut buf = String::new();
    input.read_to_string(&mut buf).unwrap();
    let mut iter = buf.trim().split_whitespace();
    let n : usize = iter.next().unwrap().parse().unwrap();
    let m : usize = iter.next().unwrap().parse().unwrap();
    
    let mut edges : Vec<Edge> = vec![];
    let mut graph : Vec<Vec<usize>> = vec![vec![]; n];
    
    for _i in 0..m {
        let from  : usize = iter.next().unwrap().parse().unwrap();
        let to    : usize = iter.next().unwrap().parse().unwrap();
        let cap   : i64 = iter.next().unwrap().parse().unwrap();
        let price : i64 = iter.next().unwrap().parse().unwrap();
        edges.push(Edge { from  : from - 1, 
                          to    : to - 1, 
                          cap   : cap, 
                          flow  : 0, 
                          price : price, 
                          rev   : 0 });
        edges.push(Edge { from  : to - 1, 
                          to    : from - 1, 
                          cap   : 0, 
                          flow  : 0, 
                          price : -price,
                          rev   : 0 });
        let len : usize = edges.len();
        edges[len - 1].rev = len - 2;
        edges[len - 2].rev = len - 1;
        graph[edges[len - 1].from].push(len - 1);
        graph[edges[len - 2].from].push(len - 2);
    }
    
    let mut cost : i64 = 0;
    
    let inf : i64 = 1_000_000_000_000;
    
    /*let mut fi : Vec<i64> = vec![inf; n]; //Djonson's potential
    fi[0] = 0;
    
    loop {
        let mut any : bool = false;
        for (num, edge) in edges.iter().enumerate() {
            if fi[edge.from] < inf 
            && edge.flow < edge.cap 
            && fi[edge.to] > fi[edge.from] + edge.price {
                fi[edge.to] = fi[edge.from] + edge.price;
                any = true;
            }
        }
        
        if !any {
            break;
        }
    }*/
    
    let mut d : Vec<i64> = vec![inf; n];
    let mut p : Vec<usize> = vec![inf as usize; n];
    let mut i : Vec<u16> = vec![0; n];
    let mut q : VecDeque<usize> = VecDeque::new();
    
    loop {
        for el in &mut d {
            *el = inf;
        }
        for el in &mut p {
            *el = inf as usize;
        }
        for el in &mut i {
            *el = 0;
        }
        d[0] = 0;
        q.push_back(0);
        
        while !q.is_empty() {
            let v : usize = q.pop_front().unwrap();
            i[v] = 1;
            for u in &graph[v] {
                let edge : &Edge = &edges[*u];
                if d[v] < inf
                && d[edge.to] > d[v] + edge.price
                && edge.flow < edge.cap {
                    d[edge.to] = d[v] + edge.price;
                    p[edge.to] = *u;
                    if i[edge.to] == 0 {
                        q.push_back(edge.to);
                    } else if i[edge.to] == 1 {
                        q.push_front(edge.to);
                    }
                    i[edge.to] = 1;
                }
            }
        }
        
        if d[n - 1] == inf {
            break;
        } else {
            let mut cur : usize = n - 1;
            let mut min : i64 = inf;
            while p[cur] != inf as usize {
                min = min.min(edges[p[cur]].cap - edges[p[cur]].flow);
                cur = edges[p[cur]].from;
            }
            cur = n - 1;
            while p[cur] != inf as usize {
                edges[p[cur]].flow += min;
                let rev = edges[p[cur]].rev;
                edges[rev].flow -= min;
                cost += min * edges[p[cur]].price;
                cur = edges[p[cur]].from;
            }
        }
    }
    
    output.write(cost.to_string().as_bytes()).unwrap();
}
